File: apps/backend/src/account/AccountService.ts
import { Injectable } from '@nestjs/common';
import {
  CreateAccountDTO,
  GetAccountDTO,
  UpdateAccountDTO,
  UUID,
} from '@mid/shared';
import { AccountRepository } from './AccountRepository';
import { AccountTypeRepository } from './AccountTypeRepository';
import { Account } from './Account';

@Injectable()
export class AccountService {
  constructor(
    private readonly accountRepository: AccountRepository,
    private readonly accountTypeRepository: AccountTypeRepository,
  ) {}

  // 계좌 생성
  async createAccount(
    userId: UUID,
    createAccountDTO: CreateAccountDTO,
  ): Promise<GetAccountDTO> {
    const { accountTypeId, ...accountData } = createAccountDTO;

    const accountType = await this.accountTypeRepository.findOneByOrFail({
      id: accountTypeId,
    });

    // Account 엔티티 생성 및 저장
    const account = this.accountRepository.create({
      ...accountData,
      accountType,
      user: { id: userId },
    });

    return await this.accountRepository
      .save(account)
      .then((account) => AccountService.toGetAccountDTO(account));
  }

  async getAccounts(userId): Promise<GetAccountDTO[]> {
    const accounts = await this.accountRepository.find({
      where: { user: { id: userId } },
      relations: ['accountType'],
    });
    return accounts.map((account) => AccountService.toGetAccountDTO(account));
  }

  async getAccountById(userId: UUID, accountId: UUID): Promise<GetAccountDTO> {
    const account = await this.accountRepository.findOneOrFail({
      where: { id: accountId, user: { id: userId } },
      relations: ['accountType'],
    });
    return AccountService.toGetAccountDTO(account);
  }

  async updateAccount(
    userId: UUID,
    accountId: UUID,
    updateAccountDTO: UpdateAccountDTO,
  ): Promise<GetAccountDTO> {
    const { accountTypeId, ...updateData } = updateAccountDTO;

    // 기존 계좌를 조회 (accountType 관계 로드 포함)
    const account = await this.accountRepository.findOneOrFail({
      where: { id: accountId, user: { id: userId } },
      relations: ['accountType'],
    });

    // 업데이트할 데이터 병합
    Object.assign(account, updateData);

    // accountTypeId가 존재할 경우 accountType을 조회하여 추가
    if (accountTypeId && account.accountType.id !== accountTypeId) {
      account.accountType = await this.accountTypeRepository.findOneOrFail({
        where: { id: accountTypeId },
      });
    }
    const updatedAccount = await this.accountRepository.save(account);

    return AccountService.toGetAccountDTO(updatedAccount);
  }

  // 계좌 삭제
  async deleteAccount(userId: string, accountId: UUID): Promise<void> {
    await this.accountRepository.deleteOrFail({
      user: { id: userId },
      id: accountId,
    });
  }

  private static toGetAccountDTO(account: Account) {
    return {
      id: account.id,
      name: account.name,
      description: account.description,
      issuer: account.issuer,
      number: account.number,
      interestRate: account.interestRate,
      withdrawalLimit: account.withdrawalLimit,
      accountTypeName: account.accountType.name,
      createdAt: account.createdAt,
    };
  }
}

File: backend/src/main/kotlin/com/kyoulho/mid/account/svc/AccountService.kt
package com.kyoulho.mid.account.svc

import com.kyoulho.mid.account.dto.CreateAccountDTO
import com.kyoulho.mid.account.dto.GetAccountDTO
import com.kyoulho.mid.account.dto.UpdateAccountDTO
import com.kyoulho.mid.account.dto.toGetAccountDTO
import com.kyoulho.mid.account.entity.Account
import com.kyoulho.mid.account.repo.AccountRepository
import com.kyoulho.mid.account.repo.AccountTypeRepository
import com.kyoulho.mid.user.repo.UserRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException

@Service
class AccountService(
    private val accountRepository: AccountRepository,
    private val accountTypeRepository: AccountTypeRepository,
    private val userRepository: UserRepository,
) {

    // 계좌 생성
    @Transactional
    fun createAccount(userId: String, createAccountDTO: CreateAccountDTO): GetAccountDTO =
        with(createAccountDTO) {
            val accountType = accountTypeRepository.findById(accountTypeId)
                .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "AccountType not found") }

            val user = userRepository.findById(userId)
                .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "MidUser not found") }

            accountRepository.save(
                Account(
                    name = name,
                    description = description,
                    issuer = issuer,
                    number = number,
                    interestRate = interestRate,
                    withdrawalLimit = withdrawalLimit,
                    accountType = accountType,
                    user = user,
                )
            ).toGetAccountDTO()
        }

    // 모든 계좌 조회
    fun getAccounts(userId: String): List<GetAccountDTO> =
        accountRepository.findByUserId(userId).map { it.toGetAccountDTO() }

    // 특정 계좌 조회
    fun getAccountById(userId: String, accountId: String): GetAccountDTO =
        accountRepository.findByIdAndUserId(accountId, userId)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Account not found") }
            .toGetAccountDTO()

    // 계좌 업데이트
    @Transactional
    fun updateAccount(userId: String, accountId: String, updateAccountDTO: UpdateAccountDTO): GetAccountDTO {
        val account = accountRepository.findByIdAndUserId(accountId, userId)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Account not found") }

        account.apply {
            if (accountType.id != updateAccountDTO.accountTypeId) {
                accountTypeRepository.findById(updateAccountDTO.accountTypeId)
                    .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "AccountType not found") }
                    .let { accountType = it }
            }
            name = updateAccountDTO.name
            description = updateAccountDTO.description
            issuer = updateAccountDTO.issuer
            number = updateAccountDTO.number
            interestRate = updateAccountDTO.interestRate
            withdrawalLimit = updateAccountDTO.withdrawalLimit
        }

        return accountRepository.save(account).toGetAccountDTO()
    }

    // 계좌 삭제
    @Transactional
    fun deleteAccount(userId: String, accountId: String) {
        accountRepository.findByIdAndUserId(accountId, userId)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Account not found") }
            .let { accountRepository.delete(it) }
    }
}
